src/main/java/net/minecraftforge/registries/DeferredRegister.java

```diff
@@ -162,17 +263,59 @@ public void handleEvent(RegistryEvent.Register<?> event) {
         return entriesView;
     }
 
+    /**
+     * @return The registry name stored in this deferred register. Useful for creating new deferred registers based on an existing one.
+     */
+    @Nullable
+    public ResourceLocation getRegistryName()
+    {
+        return this.registryName;
+    }
+
+    private Supplier<IForgeRegistry<T>> makeRegistry(final ResourceLocation registryName, final Class<T> superType, final Supplier<RegistryBuilder<T>> sup) {
+        if (registryName == null)
+            throw new IllegalStateException("Cannot create a registry without specifying a registry name");
+        if (superType == null)
+            throw new IllegalStateException("Cannot create a registry without specifying a base type");
+        if (this.type != null || this.registryFactory != null)
+            throw new IllegalStateException("Cannot create a registry for a type that already exists");
+
+        this.registryFactory = () -> sup.get().setName(registryName).setType(superType);
+        return this::getForgeRegistry;
+    }
+
+    private IForgeRegistry<T> getForgeRegistry()
+    {
+        if (this.type == null && this.typeSupplier != null)
+            this.type = this.typeSupplier.get();
+
+        return this.type;
+    }
+
+    private void onFill(IForgeRegistry<T> registry)
+    {
+        if (this.optionalTags == null)
+            return;
+
+        ITagManager<T> tagManager = registry.tags();
+        if (tagManager == null)
+            throw new IllegalStateException("The forge registry " + registry.getRegistryName() + " does not support tags, but optional tags were registered!");
+
+        Multimaps.asMap(this.optionalTags).forEach(tagManager::addOptionalTagDefaults);
+    }
+
     private void addEntries(RegistryEvent.Register<?> event)
     {
-        if (this.type == null && this.registryFactory == null)
+        IForgeRegistry<T> storedType = this.getForgeRegistry();
+        if (storedType == null && this.registryFactory == null)
         {
             //If there is no type yet and we don't have a registry factory, attempt to capture the registry
             //Note: This will only ever get run on the first registry event, as after that time,
             // the type will no longer be null. This is needed here rather than during the NewRegistry event
             // to ensure that mods can properly use deferred registers for custom registries added by other mods
             captureRegistry();
```

Comment:
This needs to be adjusted as `captureRegistry` captures the registry to `this.type`, but as references lower in this method use `storedType` without first updating storedType to point to the updated `this.type` after capturing is done, then `storedType` is still going to be null.